### ISSUE TITLE
ArgumentException.Message and ParamName

### ISSUE THREAD

**Issue by dozer75** [dotnet/coreclr#18430]
When an execption of type System.ArgumentException (or any inheritors) is thrown with paramName set it appends the text "Parameter Name: <paramName>" to the ArgumentException.Message (Ref [this](https://github.com/dotnet/coreclr/blob/1b32b2b795d52bd4bc00fd37efd528ef56b9d7af/src/System.Private.CoreLib/shared/System/ArgumentException.cs#L84)).

My opinion is that this "Parameter Name" addition has nothing to do in the Message property, that one should only contain what's . If people want this, it is possible to retrieve this using the ParamName property. Stripping it away when reading the exception is not an option since this text is localizable (e.g. "Parameternavn" in Norwegian), and thus impossible to extract in all translation variants.

So, it should be removed from ArgumentException.Message, however. It should be added to .ToString() override, since the intention of this is to display the whole argument exception information including all parameters.

If someone really requires parameter name as a part of the ArgumentExeption.Message, they should append this themselves when creating the exception.

I know that this is legacy from before .NET Core and exists in .NET Framework but I can't see why it couldn't be fixed. I suspect that this code is more copy/pasted from the .NET Framework source.

**Comment by danmosemsft** [comment]
@dozer75 could you clarify why you prefer it not be in the message? I am not clear from above. The purpose of the message in my mind  is to quickly indicate WHAT  is wrong. (For this reason we have added the file path in many IO exception messages that lacked it.)  The ToString() is mainly to indicate also WHERE it went wrong, and any nested exceptions.

I do think it is odd that it is on a different line than the rest of the message and that perhaps could change.

**Comment by danmosemsft** [comment]
>  It's not the frameworks responsibility to know what an exception should say.

It is if the exception was not created with a message in the first place, of course. As you point out, it's doing it in both cases. 

> , I can't really see WHY it should be a part of the message since it is a separate property in the ArgumentException type. Any users that want it could more easily retrieve it from there and if they want it in their message, interpolate it themselves.

Very often exceptions are found in logs or telemetry, not a debugger. The issue is long gone and maybe on a customer or remote machine. In those cases, every piece of information is valuable to help solve the problem. It is not possible to go back and examine the object fields. Generally it is better to err on the side of messy/verbose than miss something important.

> Now I have to parse the real message out without the parameter out from the message. But since the text related to the parameter is localizable, it's no way for me to remove it since the fixated (Property name) text that I can check is different in different languages.

Parsing exception messages is fragile. Right now we do not localize them, but we may do in future-  as .NET Framework does. And we can also change the messages - we don't consider that a breaking change. If your code needs some data out of the exception can it pass it elsewhere eg in a field on a derived exception or in the Data dictionary?

**Comment by danmosemsft** [comment]
I'm going to close based on the rationale above. Feel free to comment/reopen if you wish.

**Comment by dozer75** [comment]
> Very often exceptions are found in logs or telemetry, not a debugger. The issue is long gone and maybe on a customer or remote machine. In those cases, every piece of information is valuable to help solve the problem. It is not possible to go back and examine the object fields. Generally it is better to err on the side of messy/verbose than miss something important.

Well, AFAIK all log/telemetry systems I have used uses `Exception.ToString()` to retrieve the the exception information (To get all information relevant for the exception as message, stacktrace and other parameters).  In my opinion a logger/telemetry without a stacktrace is pretty much useless without a stacktrace in a program bigger than "Hello World". As I stated initially; the `ToString()` should contain the param information, that's no discussion. However the Message should always be the same as someone gave in an input is my opinion here. If a exception that inherits `ArgumentException` is in need of havning it's param name in the message, I think that it is that exception's responsibility to do it. In addition, most of the exceptions I have seen both in the framework and other places around ArgumentException actually has the param name in their message in addition to the param like `input must be set to this value.` which with the param name issue will look like:

    input must be set to this value.
    Parameter Name: input

So the parameter name is actually named twice in the message.

Generally I think it is a bad practice that when the output of a property differs from the input given, a property (like a field) should always reflect the state of the input in a get/set environment (as long as a setter exists and is used).

> Parsing exception messages is fragile. Right now we do not localize them, but we may do in future- as .NET Framework does. And we can also change the messages - we don't consider that a breaking change. If your code needs some data out of the exception can it pass it elsewhere eg in a field on a derived exception or in the Data dictionary?

I agree and this is EXACTLY why I would like the `Parameter Name:` to be removed, because in Norwegian it is `Parameternavn:` and thus the parsing doesn't work to remove this value.

Using a data dictionary is of course a solution, but then the message will be swollowed in a `ToString()` circumstance since `ToString()` doesn't parse data dictionary.

Using derived exceptions in argument handling IS actually not recommended even in the Best Practices for exceptions in Microsoft docs [Use the predefined .NET exception types](https://docs.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions#use-the-predefined-net-exception-types), and best practices is always something I want to enforce.
